kvfollowerreadsccl: make PREPARE work with bounded staleness

otan opened the pull request:
Prepared statements now work with bounded staleness reads.

In tests, we need to force trace statements for PREPARE as the EXECUTE
stmt shows up as the prepared statement during tracing.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/69027)
<!-- Reviewable:end -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 4 of 4 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69027#-:-MhMUOHMA8hs8_gHAzK2:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @otan and @rytaft)

___
*[pkg/ccl/kvccl/kvfollowerreadsccl/boundedstaleness_test.go, line 211 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/69027#-MhMTrJp0FGBGOKRHVeb:-MhMTrJq97fJf4RR7Gf3:b7jq6ct) ([raw file](https://github.com/cockroachdb/cockroach/blob/68444b521491b71ad8534c5a83d169fcf9d6cac3/pkg/ccl/kvccl/kvfollowerreadsccl/boundedstaleness_test.go#L211)):*
> ```Go
> 			// set-trace forces the next trace of events to only look for the given query
> 			// instead of using d.Input. This is useful for traces of prepared statements.
> 			case "set-trace":
> ```

Consider naming this something more specific, like `override-matching-stmt-for-tracing`. It wasn't immediately clear to me what was going on here and took some reading of the code. Consider also pointing to `boundedStalenessEvents.onStmtTrace`.
___
*[pkg/ccl/logictestccl/testdata/logic_test/as_of, line 293 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/69027#-MhMSpKoDIyqEKVaoZmI:-MhMSpKoDIyqEKVaoZmJ:b-drik1r) ([raw file](https://github.com/cockroachdb/cockroach/blob/68444b521491b71ad8534c5a83d169fcf9d6cac3/pkg/ccl/logictestccl/testdata/logic_test/as_of#L293)):*
> ```
> 
> statement ok
> PREPARE with_max_staleness_prep AS SELECT * FROM t AS OF SYSTEM TIME with_max_staleness('1ms')
> ```

Did you want to execute these first two? If not, are they redundant?


<!-- Sent from Reviewable.io -->

otan committed:
kvfollowerreadsccl: make PREPARE work with bounded staleness

Prepared statements now work with bounded staleness reads.

In tests, we need to force trace statements for PREPARE as the EXECUTE
stmt shows up as the prepared statement during tracing.

Release note: None

otan commented:
bors r=nvanbenschoten
